### ISSUE TITLE
Security audit

### ISSUE THREAD

**Issue by rugk** [borgbackup/borg#2481]
It would be nice if this backup tool could get a professional security audit. I don't know whether attic once got one, but nevertheless as much things changed in borg a security audit for it would also be a nice idea IMHO.

Maybe make a crowdfund campaign or so to raise the money and maybe also make borg popular… :smile:

**Comment by ThomasWaldmann** [comment]
Yup, maybe when 1.2 goes beta would be a good time as crypto and multithreading changes are planned.

**Comment by rugk** [comment]
How do you plan to finance it?

**Comment by enkore** [comment]
Audits are quite expensive and can cover very different levels. Some audits just look at specification or design documents (which largely don't exist for Borg), while you are probably thinking of some folks poking the code base. Given the cost there is also the latent question whether it's worth it, or if Borg is maybe not the correct target. E.g. we use msgpack a lot, which has neither been fuzzed nor audited to the best of my knowledge. Auditing Borg but then having, say, holes in msgpack would be a lot of effort for nothing :)

In the meantime this might be an interesting read: https://borgbackup.readthedocs.io/en/latest/internals/security.html

**Comment by rugk** [comment]
IMHO it's also important to check borg itself. I think some audits also cover the dependents of the projects, at least sometimes.
Also, nowadays there are many actors sponsoring audits for FLOSS software, such as [Mozilla](https://www.mozilla.org/en-US/moss/), Google or the [European](https://joinup.ec.europa.eu/community/osor/news/eu-eur-1-million-security-audit-open-source) [Union](https://juliareda.eu/2016/10/ep-votes-to-extend-fossa/).
If borg would be a ransomware it would likely get these audits for free, easily… :wink:

**Comment by enkore** [comment]
I'm afraid Borg is a bit too heavy in deployment to be used effectively for ransomware 😉

**Comment by FabioPedretti** [comment]
This is an interesting read: https://guidovranken.wordpress.com/
also https://guidovranken.wordpress.com/2017/07/06/which-software-should-i-audit-next/ (probably not relevant for borg, which has few C/C++ code).

**Comment by rugk** [comment]
And just as we're talking [GitHub also announced something](https://github.com/blog/2403-helping-secure-foss-and-the-internet-our-100-000-donation-to-the-internet-bug-bounty). The problem of course is, borg does not belong to critical infrastructure for the whole web (yet…), so make sure big comapnies start to use it as their backup tool… :wink:

**Comment by awgcooper** [comment]
I'm writing this on the assumption that a commercial security audit costs ~ $5k-10k. Subject to my precise financial circumstances at the time, I'd be willing to proffer an 'anchor amount' (~ $1,000) to the cost of an audit. One could then poll the userbase to see if the remaining amount could be crowdfunded.

Prior to doing this, I think it would make sense to:
(i) create a broad brush scope for the audit (e.g. enkore's 5 May 2017 comment above)
(ii) reach out to one or two audit companies to get a more informed view on pricing (I don't know who they are but I presume this would be easy to find out)

Does anyone have a rough idea as to how many regular users there are of borg? I have to believe it's core software infrastructure for a meaningful number of people ('define meaningful', I know). It seems to me that the helium milestone would be a good juncture to get an audit.

**Comment by Best-HeyGman** [comment]
Hello, I would be willing to add 500€ to the audit fund.

The security audit for gocryptfs was done by Defuse Security (https://defuse.ca/software-security-auditing.htm) and it seems that they have experience in auditing open source software.

**Comment by Best-HeyGman** [comment]
Also, I do think a good time for the audit would be before Borg 2.0.0 stable is released, as it would be very difficult to make changes to the security architecture after that point.

**Comment by ThomasWaldmann** [comment]
@Best-HeyGman Thanks for offering your support!

I agree that before borg2 stable release would be a good time.

**Comment by Best-HeyGman** [comment]
@ThomasWaldmann I see you've added this to the 2.0.0 Milestone
Is it ok for you if I put my 500€ in via Open Collective? (https://opencollective.com/borgbackup)

**Comment by ThomasWaldmann** [comment]
@Best-HeyGman IIRC (didn't do much yet with OC): the funds collected via Open Collective are not bound to specific issues like "doing a security audit".

Thanks for intending to support this!

**Comment by Best-HeyGman** [comment]
@ThomasWaldmann Yeah, I guess a better way would be setting up some kind of gofundme after getting a quote on how much money is needed.

However, I didn't want to burden you with having to set that up when the Open Collective is already there.

**Comment by Best-HeyGman** [comment]
@ThomasWaldmann 
Alright, here it is, before I forget about this again: https://opencollective.com/borgbackup/contributions/865259
Don't worry, if there's no security audit in the end, just keep it. Borgbackup deserves it either way :)

**Comment by ThomasWaldmann** [comment]
@Best-HeyGman Thanks a lot for supporting the project!

**Comment by ThomasWaldmann** [comment]
Considering we do not have a lot of funds for a paid potentially very expensive human security audit, this is an interesting development:

https://www.heise.de/en/news/AI-model-Claude-detects-Firefox-bugs-faster-than-the-community-11202815.html

**Comment by ThomasWaldmann** [comment]
AI already found some smaller issues. \o/

**Comment by awgcooper** [comment]
I understand that the Borg 2 code base has involved some fundamental changes and has added more lines to the total. If you think "AI" will be sufficient that is great. However if, after having used "AI" to the fullest extent possible you believe that a 'very expensive' human audit is still the right thing to do, don't hesitate to shout out! We will find the funds.
